DT tooling and DT API for RANGES property

superna9999 opened the pull request:
As requested in https://github.com/zephyrproject-rtos/zephyr/pull/37668#pullrequestreview-729799792, this adds support for parsing and extracting the DT RANGES property with the associated macros.

The ranges property is used standalone in PCIe DT bindings, As described in IEEE Std 1275-1994, to describe the PCI I/O and memory regions.

superna9999 committed:
edtlib: add Range property on Node, used for PCIe I/O and memory regions

As described in IEEE Std 1275-1994, the PCIe bindings uses the ranges
property to describe the PCI I/O and memory regions.

Add parsing of this property in edtlib.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
gen_defines: generate ranges defines for PCIe I/O and memory regions

As described in IEEE Std 1275-1994, the PCIe bindings uses the ranges
property to describe the PCI I/O and memory regions.

Write _RANGES_ defines that will be used to determines the I/O and
memory regions from PCIe Controller drivers.

Also exclude "ranges" & "dma-ranges" property's length generation
alogn "reg" and "interrupt".

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
devicetree: add DT ranges public API

This adds the macro public API to handle the DT ranges properties.

This also updates the devicetree/api.rst documentation.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 commented:
Seems doctree doesn't like #address-cells and #size-cells in doc, is there a way to do this correctly ?

mbolivar-nordic commented:
> Seems doctree doesn't like #address-cells and #size-cells in doc, is there a way to do this correctly ?

You need to indent the code examples so doxygen recognizes them as preformatted code blocks. See the other headers for examples.

superna9999 commented:
@galak is the API ok so I can add tests in order to merge #37688 ?

superna9999 committed:
devicetree: add tests for devicetree ranges API macros

Test cases for new APIs.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

mbolivar-nordic reviewed (changes requested):
temporary -1 until I've had a chance to review

mbolivar-nordic commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1131,6 +1132,64 @@ def _check_undeclared_props(self):
+            if child_address_cells == 0:
+                range_.child_bus_flags = None
+                range_.child_bus_addr = None
+            elif child_address_cells == 3:
+                # When used as PCIe ranges, the first entry can be some flags
```
This seems to be special case code for PCIe, but there is no indication that the node is related to PCIe, so why is this here? I don't understand how it's safe to assume this in general.; nothing in the DT spec says that 3 is a magic value which means the first entry is always PCIe-specific flags, and this code runs for every node in the DT.

superna9999 replied:
(in ref to https://github.com/zephyrproject-rtos/zephyr/pull/38106#discussion_r718295619)
Yep I understand, but I don't see any other way to handle `#address-cells = 3`

mbolivar-nordic replied:
>  Yep I understand, but I don't see any other way to handle `#address-cells = 3`

Is there some other way the node marks itself as a PCIe controller? Some sort of property or node name convention?

mbolivar-nordic replied:
... and if not, perhaps we can get some extra hints from the bindings to use in edtlib. E.g. we can treat the node differently if and only if we see `bus: pci` (https://docs.zephyrproject.org/latest/guides/dts/bindings.html#bus) in its matching binding.

superna9999 replied:
@mbolivar-nordic OK I did that.

I only strictly parsed Ranges in edtlib (so address-cells generates 92bit numbers in python, but it's ok) and fixed the tests accordingly. I also passed the cells count for each part of a range, because it's used in the case of PCIe to know the width of the first range entry to extract the flags.

Then I simply used the bus bindings property in gen_defines to extract the flags.

mbolivar-nordic commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1131,6 +1132,64 @@ def _check_undeclared_props(self):
+
+            self.ranges.append(range_)
+
+        _add_names(node, "ranges", self.ranges)
```
I'm not familiar with names being used for the ranges property; can you please provide a reference to explain what's going on here?

superna9999 replied:
Isn't `add_names` necessary to be used from `gen_defines` ?

mbolivar-nordic reviewed (changes requested):
Some initial comments on the edtlib.py changes. Thanks for your patience!

superna9999 commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1294,6 +1353,39 @@ def _named_cells(self, controller, data, basename):
+    These attributes are available on Range objects:
+
+    node:
+      The Node instance this register is from
```
yep, will fix

superna9999 commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1294,6 +1353,39 @@ def _named_cells(self, controller, data, basename):
+    node:
+      The Node instance this register is from
+
+    child_bus_flags:
+      Represents a bitfield for the translation, commonly used in PCI mapping
```
Ranges in PCI is not obvious.

The PCI controller nodes uses the standard ranges property to declare memory ranges to be used with childs.

But these are not used directly, they are negociated with the BAR child pci registers, so translation is not to be done like the classic DT bus ranges.

Basically it uses the standard device tree ranges (as defined in https://buildmedia.readthedocs.org/media/pdf/devicetree-specification/latest/devicetree-specification.pdf pages 18/29) encoded as an arbitrary number of (`child-bus-address`
, `parent-bus-address`, `length`) triplets.

But the first `child-bus-address` has a different `#address-cells`, 2 or 3 depending on the parent node #address-cells`.
The extra cell encodes the memory ranges properties, this was defined a long time ago in IEEE Std 1275-1994 (https://www.openfirmware.info/data/docs/bus.pci.pdf page 4).

The main problem is that today the ranges cells are interpreted as C numbers in headers, with `#address-cells = 2` it's ok and the flags cell can be extracted in C, but with `#address-cells = 3` it's another issue.

I considered that `#address-cells = 3` is unique to PCI, and so splitting the extra cell into a separate `child_bus_flags` attribute was simpler.

mbolivar-nordic replied:
Thanks for the reference; I will read through the IEEE doc.

superna9999 replied:
For reference here is how u-boot handles the ranges attribute: https://github.com/u-boot/u-boot/blob/b5d7a200a81097fddef7ba5ec42e330d49fa2f33/drivers/pci/pci-uclass.c#L960

superna9999 replied:
Here for Linux https://github.com/torvalds/linux/blob/02d5e016800d082058b3d3b7c3ede136cdc6ddcb/drivers/pci/of.c#L274

superna9999 replied:
The main difference it they handle the ranges node dynamically because they know the node is a pci node

superna9999 replied:
OK, I dropped this from edtlib, see https://github.com/zephyrproject-rtos/zephyr/pull/38106#discussion_r719232883

superna9999 commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1294,6 +1353,39 @@ def _named_cells(self, controller, data, basename):
 
+class Range:
+    """
+    Represents a translation range on a node.
```
ok

superna9999 commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1294,6 +1353,39 @@ def _named_cells(self, controller, data, basename):
+      Represents a bitfield for the translation, commonly used in PCI mapping
+
+    child_bus_addr:
+      Is a physical address within the child bus address space
```
ok

superna9999 commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1131,6 +1132,64 @@ def _check_undeclared_props(self):
+        # Initializes self.ranges
+        node = self._node
+
+        self.ranges = []
```
ok will add

superna9999 commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1131,6 +1132,64 @@ def _check_undeclared_props(self):
+        else:
+            child_address_cells = child_address_cells.to_num()
+        child_size_cells = node.props.get("#size-cells")
+        if not child_size_cells:
```
ok

superna9999 commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1131,6 +1132,64 @@ def _check_undeclared_props(self):
+                .format(child_address_cells, parent_address_cells,
+                    child_size_cells)):
+
+            range_ = Range()
```
My bad, Vim detected `range` as a reserved keyword... will remove.

superna9999 commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1131,6 +1132,64 @@ def _check_undeclared_props(self):
+                "4*(<#address-cells> (= {}) + "
+                "<#address-cells for parent> (= {}) + "
+                "<#size-cells> (= {}))"
+                .format(child_address_cells, parent_address_cells,
```
ok

superna9999 commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1131,6 +1132,64 @@ def _check_undeclared_props(self):
+
+        child_address_cells = node.props.get("#address-cells")
+        parent_address_cells = _address_cells(node)
+        if not child_address_cells:
```
ok

superna9999 committed:
edtlib: add tests to check Range property

This adds some tests in test_edtlib.py and test.dts to check all
common possible combination of ranges property usage and handling
by edtlib.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 commented:
Following the rework, I completely cleaned the `devicetree/ranges.h` header and the related tests.

mbolivar-nordic commented:
Thanks @superna9999 . I'll get back to you this week; I need some time to digest.

mbolivar-nordic reviewed (changes requested):
I'm a fan of the latest edtlib patch. Thank you.

I took some time to try to understand the IEEE document as well and I think I have a rough idea now. Thanks again for the references.

Here are some more comments on the rest of the PR. I haven't looked at the test cases in detail yet, but this is what I'm thinking on the other unreviewed patches.

superna9999 committed:
doc: document ranges generated macros in macros.bnf

This adds the DT ranges generated macros in the ABNF grammar file.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
devicetree: update DT_PROP_LEN documentation about ranges and dma-ranges

Like reg and interrupts, using DT_PROP_LEN doesn't make sense since
the ranges & dma-ranges block lengths and values depends on the DT
node #address-cells, #size-cells and parents #address-cells value.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

mbolivar-nordic reviewed (changes requested):
Looking good.

I found some typos and things in the docstrings, but in general I totally support the direction you took with where the macros go and how you are documenting them.

I will take a look at the tests tomorrow. After those are reviewed, I'll be ready to approve.

mbolivar-nordic reviewed (changes requested):
Test cases look great. Very thorough :).

I ended up having some style nitpicks and things that were easier to explain as a commit on top:

https://github.com/mbolivar-nordic/zephyr/commit/3b7e0a8341d1dd610f04783c1497b39d8fed98b4

Could you please take that patch? Then I think I'm ready to approve.

What do you think about the stable API change process? Did you want me to handle that for you, or did you want to do it yourself?

Here's a direct link to the steps in the process: https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html#introducing-incompatible-changes

Almost there! :slightly_smiling_face:

mbolivar-nordic commented:
> I'll try to start the stable API process, but, should I start it for this PR or should I open a separate PR for [9685ca8](https://github.com/zephyrproject-rtos/zephyr/commit/9685ca867640baa6a8380c87d5fc5212a3951950) ?
> 
> And do I need to also do the same process for [da0cfd6](https://github.com/zephyrproject-rtos/zephyr/commit/da0cfd618365f0d328a809714139595b33d23b61) ?

I think that in this case, it is OK to do a single stable API change process for both commits, since they are part of the same PR, which is making one logical change ("use DT_NUM_RANGES instead of DT_PROP_LEN for the ranges property") to the API.

mbolivar-nordic reviewed (approved):
Approving after marking DNM / Stable API Change / etc.

@superna9999, you just need to follow the process now. If nobody objects at the meeting where this is decided, you should be fine.

@nashif recently started discussion at the TSC about moving stable API changes to that forum from the API meeting, so I'm not sure what meeting you'll need to discuss it at, but that won't stop you from getting the process going.

galak reviewed (commented):
What are we doing about dma-ranges?

mbolivar-nordic commented:
> The minimal should be to exclude `dma-ranges` like `interrupt`, `reg` and `ranges` like in:

My opinion is that this is acceptable to include this while we're breaking things for regular `ranges`, and then defer `dma-ranges` support to potential future work.

If you do, you will need to update the stable API change request etc.

galak reviewed (commented):
So let's add `dma-ranges` to the comment now so we don't need to do that in the future.  I don't think this requires going to the API meeting since nothing uses or supports `dma-ranges` so no issue.

galak commented:
We should probably update release-notes as part of this PR.

superna9999 commented:
@galak @mbolivar-nordic done

superna9999 commented:
@mbolivar-nordic @galak Ping.

Are the `release-notes` and `dma-ranges` changes ok ?

galak commented:
> @mbolivar-nordic @galak Ping.
> 
> Are the `release-notes` and `dma-ranges` changes ok ?

LGTM.

galak commented:
Looks like a merge conflict in release notes.

superna9999 commented:
@galak Fixed
